### ISSUE TITLE
`docker network create` is not atomic

### ISSUE THREAD

**Issue by thanatos** [docker/cli#614]
**Description**

The command `docker network create`, generally speaking, does not allow duplicate networks with the same name; however, multiple invocations of the command, running in parallel, can successfully create separate networks with the same name.

We are using the network label as a means to uniquely identify which network new containers should be attached to. The code that is launching the containers, will, if it detects the network doesn't exist, attempt to create it. During deployment, because all the containers on the VM are attempt to start simultaneously, they appear to race the network creation, and end up creating multiple networks.

**Steps to reproduce the issue:**
1. Run `docker network create foo` multiple times in parallel (this happens quite reliably on my machine, despite seeming to be a race condition):

   ```bash
   » docker network create foo & docker network create foo
   [1] 37777
   84db9b0ee0e0db89d6d2636a7cd89961c44018d9fb656da4933b311bb2f6f976
   [1]  + done       docker network create foo
   801bbe64794a2bf26055c24b307635978fda484c8d1ec88cb8baf39ad4e18cff
   » docker network ls
   NETWORK ID          NAME                    DRIVER              SCOPE
   80c6db7c769e        bridge                  bridge              local
   84db9b0ee0e0        foo                     bridge              local
   801bbe64794a        foo                     bridge              local
   3e7fcdba9fa1        host                    host                local
   0110ad46f41d        none                    null                local
   » docker network create foo
   Error response from daemon: network with name foo already exists
   ```

**Describe the results you received:**
As you can see, by running it quickly, we get two networks w/ the same name.

**Describe the results you expected:**
Since the command, when run slowly, disallows duplicates, I expect duplicates to not be possible. In general, as a user, I find duplicates problematic: having two networks with the same name makes it incredibly difficult to reason about the system as a whole. For example, what network will `docker run --network foo` use? Is it well defined? As a user, I'd also like to be able to refer to networks by label for readability, and for knowing if a network for a set of containers has been created / needs to be created. (Presently, I would need to keep track of that information externally to Docker, and also lock around network creation externally to Docker, both of which are pains that I'd rather just avoid.)

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
» docker version
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:09 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:45:38 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 3
 Running: 0
 Paused: 0
 Stopped: 3
Images: 1283
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.49-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.952GiB
Name: moby
ID: 4EAX:SZMF:N4IY:W5OM:5UR5:JNXI:7M7N:K6ZQ:NPEA:T27H:76ZB:3WAK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 31
 System Time: 2017-10-10T19:34:17.407905594Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
We also replicated this bug on Docker CE running on Linux on a AWS VM. That VM's `docker version`,
```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by dnephin** [comment]
This is not really an issue with the CLI, the behaviour is part of the API. I would suggest opening an issue on https://github.com/docker/for-linux

**Comment by thaJeztah** [comment]
This is a duplicate of https://github.com/moby/moby/issues/18864 (and https://github.com/moby/moby/issues/29268, https://github.com/moby/moby/issues/33037).

Allowing duplicate names (but unique ID's) was by design, as explained in https://github.com/moby/moby/issues/18864#issuecomment-167006094, but feel free to participate in that discussion.

Let me close the issue here, because this is an issue in the API, and not the client (which is what's maintained in this repository)
